### PR TITLE
feat: Add types

### LINKED ISSUE #18845: Change Request: Publish types
### ESLint version

HEAD

### What problem do you want to solve?

Right now, we're in a situation where there's a `@types/eslint` package that people use for type definitions for ESLint. This package is very popular (12M downloads per week), but it's always out of date because we don't maintain it. This continues to cause issues as we work on language plugins and the core rewrite.

Further, it can take weeks to get changes to `@types/eslint` approved([last PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69957)) and deployed. Because we'll likely have more type changes as we accelerate language plugins and core rewrite work, this just doesn't seem sustainable.

### What do you think is the correct solution?

ESLint should publish its own types.

I think the easiest place to start is just to copy the type definitions from `@types/eslint` as-is and publish those as part of the `eslint` package. This will be provide a good level of compatibility for those using `@types/eslint`.

We can then easily make modifications, including references to `@eslint/core` as we go forward.

In terms of types from `eslint` and `@eslint/core`:

* `@eslint/core` types will be exclusively those types that are for use in the core rewrite. This excludes API interfaces such as `ESLint` and `Linter`.
* `eslint` types will be a superset of `@eslint/core` that includes API interfaces for the `eslint` package.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Added the type definitions from `@types/eslint` to the project:

- Added all type definitions in `lib/types`
- Added tests in `tests/lib/types`
- Added new `test:types` script to run type checking tests
- Added `npm run test:types` to CI to be run on each PR
- Added `types` entries to `package.json`

For simplicity, I did not enable type checking for the whole repo, just for the type tests.

This is intended to be a base for us to build off of. We can look to make improvements in future PRs.

fixes #18845

#### Is there anything you'd like reviewers to focus on?

The types are copied over as-is without modification (aside from including the license). We don't need to review their contents right now.

Does the testing setup make sense?

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -93,7 +93,8 @@ module.exports = [
             "tests/performance/**",
             "tmp/**",
             "**/test.js",
-            ".vscode"
+            ".vscode",
+            "**/*.ts"
```
Comment (nzakas) [inline eslint.config.js:97]: Just for now. :smile:

**Comment by nzakas** [comment]
Looks like the webpack plugin is using an older version of `@types/eslint`, but still seems like there may be an error in the types for formatters?

**Comment by fasttime** [comment]
I can confirm the `Formatter` type is broken, see https://github.com/eslint/eslint/issues/18821#issuecomment-2313177061. Would patching that type help to fix the error?

**Comment by nzakas** [comment]
Yes, I think that's all it would take to make both integration tests pass. I'll give it a shot.

**Review by fasttime — APPROVED** [review]
LGTM, thanks! I'm going to merge this PR now so that it gets included in the ESLint v9.10.0 release.

**Comment by sam3k** [comment]
Per 2024-09-05 TSC meeting, this was released early and we will await feedback from users to make sure it is working as expected. 

@nzakas added some integration tests for the types.

**Comment by dimitropoulos** [comment]
thank you for doing this!

**Pull request merged**